src/algebra/char_p/mixed_char_zero.lean

```diff
@@ -0,0 +1,471 @@
+  apply Q_algebra_iff_equal_char_zero,
+end
+
+/-!
+# Splitting statements into different characteristic
+
+Statements to split a proof by characteristic. There are 3 theorems here that are very
+similar that only differ in the assumptions we can make on the positive characteristic
+case:
+Generally we need to consider all `p ≠ 0`, but if `R` is a local ring, we can assume
+that `p` is a prime power and if `R` is a domain, we can even assume that `p` is prime.
+-/
+section main_statements
+
+variable {P : Prop}
+
+/--
+Split a `Prop` in characteristic zero into equal and mixed characteristic.
+-/
+theorem split_equal_mixed_char [char_zero R]
+  (h_equal : algebra ℚ R → P)
+  (h_mixed : ∀ (p : ℕ), (nat.prime p → mixed_char_zero R p → P)) : P :=
+begin
+  by_cases h : ∃ (p : ℕ), mixed_char_zero R p,
+  { cases h with p hp,
+    rw ←mixed_char_zero.reduce_to_p_prime at h_mixed,
+    apply h_mixed p,
+    exact hp },
+  { apply h_equal,
+    rw [←not_Q_algebra_iff_not_equal_char_zero, not_is_empty_iff] at h,
+    exact h.some },
+end
+
+/-- Split any `Prop` over `R` into the three cases:
+- positive characteristic.
+- equal characteristic zero.
+- mixed characteristic `(0, p)`.
+-/
+theorem split_by_characteristic
+  (h_pos : ∀ (p : ℕ), (p ≠ 0 → char_p R p → P))
+  (h_equal : algebra ℚ R → P)
+  (h_mixed : ∀ (p : ℕ), (nat.prime p → mixed_char_zero R p → P)) : P :=
+begin
+  cases char_p.exists R with p p_char,
+  by_cases p = 0,
+  { rw h at p_char,
+    resetI, -- make `p_char : char_p R 0` an instance.
+    haveI h0 : char_zero R := char_p.char_p_to_char_zero R,
+    exact split_equal_mixed_char R h_equal h_mixed },
+  exact h_pos p h p_char,
+end
+
+/-- In a `is_domain R`, split any `Prop` over `R` into the three cases:
+- *prime* characteristic.
+- equal characteristic zero.
+- mixed characteristic `(0, p)`.
+-/
+theorem split_by_characteristic_domain [is_domain R]
+  (h_pos : ∀ (p : ℕ), (nat.prime p → char_p R p → P))
+  (h_equal : algebra ℚ R → P)
+  (h_mixed : ∀ (p : ℕ), (nat.prime p → mixed_char_zero R p → P)) : P :=
+begin
+  refine split_by_characteristic R _ h_equal h_mixed,
+  introsI p p_pos p_char,
+  have p_prime : nat.prime p := or_iff_not_imp_right.mp (char_p.char_is_prime_or_zero R p) p_pos,
+  exact h_pos p p_prime p_char,
+end
+
+/-- In a `local_ring R`, split any `Prop` over `R` into the three cases:
+- *prime power* characteristic.
+- equal characteristic zero.
+- mixed characteristic `(0, p)`.
+-/
+theorem split_by_characteristic_local_ring [local_ring R]
+  (h_pos : ∀ (p : ℕ), (is_prime_pow p → char_p R p → P))
+  (h_equal : algebra ℚ R → P)
+  (h_mixed : ∀ (p : ℕ), (nat.prime p → mixed_char_zero R p → P)) : P :=
+begin
+  refine split_by_characteristic R _ h_equal h_mixed,
+  introsI p p_pos p_char,
+  have p_ppow : is_prime_pow p := or_iff_not_imp_left.mp (char_p_zero_or_prime_power R p) p_pos,
+  exact h_pos p p_ppow p_char,
+end
+
+end main_statements
+
+/-!
```

Comment:
Not sure if this last section about the coersion `ℕ+ → Rˣ` is something that is useful to have or not...